### ISSUE TITLE
Feature request: make fuzz!() usable in unit tests

### ISSUE THREAD

**Issue by ckaran** [camshaft/bolero#8]
I'm slowly converting my code to using bolero from using [proptest](https://crates.io/crates/proptest), mainly because I think bolero has better fuzz testing.  However, I'd like to be able to use fuzz!() within unit tests (I know about check!(), but since it isn't fuzzing, I'm not as interested in it).  I know that this doesn't work right now; when I try to use fuzz!() within even simple unit tests I get a segfault.  But since I know very little about how bolero works under the hood, I don't know if this is something that is easy to fix, or if it will be a major effort.  So, would it be easy to get fuzz!() to work correctly within unit tests?  If not, can you please tell me why, just to satisfy my curiosity?

**Comment by camshaft** [comment]
Good question! I initially wanted `fuzz!` to be used everywhere but ran into a couple issues with it. The current fuzzing engines (`afl`, `libfuzzer`, `honggfuzz`) aren't initially compatible with the way unit tests are executed. The engines expect exclusive control over the event loop (`libfuzzer` is compiled into the executable itself, `afl` and `honggfuzz` communicate with an external parent executable). Fixing this would potentially require messing with Rust's default test harness, which seemed like more work that it was worth at the time. It may be worth revisiting.

Test threads need to be disabled, which I believe can be accomplished by passing `--test-threads 1` but I'd need to look more in depth to make sure.

Another problem I ran into at the time was being able to run a specific test. The default behavior when running something like `cargo test my_test` is it will execute all tests containing the string `my_test` e.g. `my_test`, `my_test_a`, and `a_my_test`. I recently discovered the `--exact` flag can be passed to fix that problem so that wouldn't be an issue.

The other issue I ran into was figuring out where to put the corpus files for the unit tests. The integration tests work out really nice:

```
tests
└── round_trip
    ├── corpus
    │   ├── 02b4a70c5a3a9fb919f7a37bdda22eac9d53c1fe
    │   └── 0741f925fb63b6f2b6d7e25179cc285ebe957f2a
    ├── crashes
    │   ├── 0b5a06f5a02e5e674d085192a21ababe5c06bf64
    │   └── 0b925d94e02a5817d9918f215fecb07bbcb0cb2c
    └── main.rs
```

I would want a file structure that doesn't litter the `src` tree with a bunch of test directories.

Anyway, those are my initial thoughts. There's probably other reasons that I'm not thinking about but can't currently recall.

Question for you: What your reason for wanting unit test support?

**Comment by ckaran** [comment]
So I can test private interfaces without having to making them public.  Following the advice in [the book](https://doc.rust-lang.org/book/ch11-03-test-organization.html) makes this easy, but I don't know how well it will work for isolated test cases such as bolero uses.

You're right about the corpus files though; I don't have a good answer for that.  Maybe recreate the `src` directory structure within the `tests` directory?  That is, if the path to a test is `crate::roundtrip::tests::test1`, then you'll have `tests/roundtrip/tests/test1/corpus` and `tests/roundtrip/tests/test1/crashes`.  Alternatively, to ensure that you don't have any namespace collisions, you could create a new top level directory that is named `bolero_unittests` that is beside the `tests` directory.  (If you do that, you might want to generate a `README.md` within `bolero_unittests` that explains that the directory is a scratch space, and could be completely wiped clean at any time.  That will prevent people from accidentally trying to store their own stuff in there.)

**Comment by camshaft** [comment]
implemented in #7

**Comment by ckaran** [comment]
Thank you!
